src/Compiler/Service/IncrementalBuild.fs

```diff
@@ -856,33 +856,34 @@ module IncrementalBuilderHelpers =
         }
 
     /// Finish up the typechecking to produce outputs for the rest of the compilation process
-    let FinalizeTypeCheckTask (tcConfig: TcConfig) tcGlobals enablePartialTypeChecking assemblyName outfile (boundModels: ImmutableArray<BoundModel>) =
+    let FinalizeTypeCheckTask (tcConfig: TcConfig) tcGlobals partialCheck assemblyName outfile (boundModels: GraphNode<BoundModel> seq) =
       node {
         let diagnosticsLogger = CompilationDiagnosticLogger("FinalizeTypeCheckTask", tcConfig.diagnosticsOptions)
         use _ = new CompilationGlobalsScope(diagnosticsLogger, BuildPhase.TypeCheck)
 
+        let! computedBoundModels = boundModels |> Seq.map (fun g -> g.GetOrComputeValue()) |> NodeCode.Parallel
```

Comment:
Is running this in parallel actually helpful? Since when the bound models are created in `createBoundModelGraphNode` they are chained together sequentially.

I know that chain _can_ be broken when implementation file back by signature is edited. But it probably doesn't happen systematically.

Reply:
A bit but not very much. It takes the parsing part from that chain and runs it independently. Basically anytime a bound model later in that sequence would wait for a parse, it doesn't, because the results are already there.
I guess most gains would be in scenarios of full typecheck or user opening the last file in a project.

 User edits are another thing as they will cause reparsing only of the file edited.

Reply:
> It takes the parsing part from that chain and runs it independently

How can it do that? The first thing that `GetOrComputeValue` does is wait for previous model's `GetOrComputeValue`. How can the parsing be triggered before that finishes?

Reply:
Because parsing is independent from typecheck, we can do it as soon as we have the sources and we have them all at the creation of InitialState.

Reply:
Ah, damn it I completely missed what you were asking about and was talking about parsing, my apologies @0101.